OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+template<typename ET = SimTK::Real> // Element datatype.
+class OpenSim::DataTable_ : public OpenSim::AbstractDataTable {
+public:
+  using size_t                     = std::size_t;
+  using string                     = std::string;
+  using MetaDataType               = std::unordered_map<string, 
+                                                        SimTK::AbstractValue*>;
+  using ColLabelsType              = std::unordered_map<string, size_t>;
+  using ColLabelsIterType          = ColLabelsType::iterator;
+  using ColLabelsConstIterType     = ColLabelsType::const_iterator;
+  using ColLabelsIterPairType      = std::pair<ColLabelsIterType,
+                                               ColLabelsIterType>;
+  using ColLabelsConstIterPairType = std::pair<ColLabelsConstIterType,
+                                               ColLabelsConstIterType>;
+
+  /// Construct empty DataTable.
+  DataTable_() :
+    data{}, metadata{}, col_ind{} {}
+
+  /// Construct DataTable of given shape populated with given value val. 
+  /// Default value for val is NaN.
+  /// \param nrows Number of rows.
+  /// \param ncols Number of columns.
+  /// \param val   Value to initialize all the entries to.
+  DataTable_(size_t nrows,
+             size_t ncols,
+             const ET& val = ET{SimTK::NaN}) :
+    data{int(nrows), int(ncols), val}, metadata{}, col_ind{} {}
+
+  /// Construct DataTable using an iterator(satisfying requirement of an 
+  /// input_iterator) which produces one entry at a time. The entries of 
+  /// DataTable_ are copy initialized using the values produced by the 
+  /// iterator. For example, specifying ROWWISE for parameter dir and 10 for 
+  /// parameter ndir will populate the DataTable one row at a time with each 
+  /// row's length taken to be 10 elements.
+  ///
+  /// \param first Beginning of range covered by the iterator.
+  /// \param last End of the range covered by the iterator.
+  /// \param ndir Extent of the dimention specified by parameter dir. 
+  /// \param dir Dimension to populate the DataTable. Populate row-wise or 
+  ///              column wise. See OpenSim::InputItDir for possible values.
+  /// \param allow_missing Allow for missing values. When set to false, this
+  ///                      function will throw if the input iterator fills up 
+  ///                      the last row/col partially. When set to true, 
+  ///                      missing elements are set to SimTK::NaN.
+  ///
+  /// \throws ZeroElements When input-iterator does not produce any elements.
+  ///                      That is first == last.
+  template<typename InputIt>
+  DataTable_(InputIt first,
+             typename std::enable_if<!std::is_integral<InputIt>::value,
+                                     InputIt>::type last,
+             const size_t ndir,
+             const InputItDir dir = ROWWISE,
+             const bool allow_missing = false) :
```

Comment:
These three arguments are pass by value and thus should not be `const` here. That's because this signature is doubling as API documentation and `const` is not meaningful from the caller's point of view. It is only meaningful for the implementer and thus must not be shown as part of the API. It is only here because you happened to put the implementation in the header file.

This is actually a bug in the C++ language (or maybe just the compiler implementations). It should have been designed so that the API could be declared with parameters like `f(int i)` and then the implementation could be `f(const int i)` recognized as the same signature, but with the implementer making it clear that they won't be modifying the argument in the implementation.